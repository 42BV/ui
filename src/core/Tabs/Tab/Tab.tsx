import React from 'react';
import { NavItem, NavLink } from 'reactstrap';
import { Icon, IconType } from '../../Icon';
import { Color } from '../../types';
import { alwaysTrue } from '../../../form/utils';
import classNames from 'classnames';

export type Props = {
  /**
   * The label to display in the tab navigation.
   */
  label: string;

  /**
   * Callback that should activate the tab.
   *
   * We highly recommend adding the tab as a URL parameter,
   * especially if the inactive tabs are not hidden by CSS.
   */
  onClick: () => void;

  /**
   * Callback to render the tab content.
   */
  children: () => React.ReactNode;

  /**
   * Whether the tab content should be displayed.
   */
  active: boolean;

  /**
   * Optional boolean to determine if the tab navigation item should be disabled.
   *
   * Defaults to false.
   */
  disabled?: boolean;

  /**
   * Optional callback to determine if the tab should be displayed.
   * This is useful if you have to do permission checks to determine
   * if the user is allowed to see the tab content.
   *
   * By default, tabs will always be displayed.
   */
  show?: () => boolean;

  /**
   * Optional icon to be displayed in front of the label in the tab navigation.
   */
  icon?: IconType;

  /**
   * Optional color the icon should have.
   * This is useful when using tabs in a form and an inactive tab
   * contains validation errors.
   */
  iconColor?: Color;

  /**
   * Optional extra CSS class you want to add to the component.
   * Useful for styling the component.
   */
  className?: string;
};

/**
 * The Tab component displays a tab navigation item and holds the content
 * that will only be displayed when the tab is active.
 * This component should only be used inside the Tabs component.
 */
export function Tab({
  icon,
  label,
  onClick,
  active,
  show = alwaysTrue,
  disabled,
  iconColor,
  className
}: Props) {
  if (!show()) {
    return null;
  }

  return (
    <NavItem className={classNames({ disabled }, className)}>
      <NavLink active={active} onClick={onClick} disabled={disabled}>
        <div className="d-flex p-2 justify-content-center">
          {icon ? (
            <Icon icon={icon} color={iconColor} className="me-1" />
          ) : null}
          {label}
        </div>
      </NavLink>
    </NavItem>
  );
}
