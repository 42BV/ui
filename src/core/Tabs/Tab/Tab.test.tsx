import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import { Tab } from './Tab';

describe('Component: Tab', () => {
  function setup({
    active = true,
    hasIcon,
    hasIconColor,
    show,
    disabled
  }: {
    active?: boolean;
    hasIcon?: boolean;
    hasIconColor?: boolean;
    show?: boolean;
    disabled?: boolean;
  }) {
    const onClickSpy = jest.fn();
    const showSpy = jest.fn().mockReturnValue(show);

    const tab = shallow(
      <Tab
        active={active}
        label="test"
        onClick={onClickSpy}
        icon={hasIcon ? 'close' : undefined}
        iconColor={hasIconColor ? 'primary' : undefined}
        show={show !== undefined ? showSpy : undefined}
        disabled={disabled}
      >
        {() => <p>test</p>}
      </Tab>
    );

    return { tab, onClickSpy, showSpy };
  }

  it('should return null when show returns false', () => {
    const { tab, showSpy } = setup({ show: false });

    expect(showSpy).toHaveBeenCalledTimes(1);
    expect(toJson(tab)).toBe('');
  });

  describe('ui', () => {
    test('without icon', () => {
      const { tab } = setup({});

      expect(toJson(tab)).toMatchSnapshot(
        'Component: Tab => ui => without icon'
      );
    });

    test('with icon', () => {
      const { tab } = setup({ hasIcon: true });

      expect(toJson(tab)).toMatchSnapshot('Component: Tab => ui => with icon');
    });

    test('icon color', () => {
      const { tab } = setup({ hasIcon: true, hasIconColor: true });

      expect(tab.find('Icon').props().color).toBe('primary');
    });

    test('not active', () => {
      const { tab } = setup({ active: false });

      // @ts-expect-error Test mock
      expect(tab.find('NavLink').props().active).toBe(false);
    });

    test('not active', () => {
      const { tab } = setup({ active: false, disabled: true });

      expect(tab.find('NavLink').props().disabled).toBe(true);
    });
  });

  describe('events', () => {
    it('should call onClick when NavLink is clicked', () => {
      const { tab, onClickSpy } = setup({});

      expect(onClickSpy).toHaveBeenCalledTimes(0);

      // @ts-expect-error Test mock
      tab.find('NavLink').props().onClick();

      expect(onClickSpy).toHaveBeenCalledTimes(1);
    });
  });
});
