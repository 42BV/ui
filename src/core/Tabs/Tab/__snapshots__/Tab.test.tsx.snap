// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Tab ui with icon: Component: Tab => ui => with icon 1`] = `
<NavItem
  className=""
  tag="li"
>
  <NavLink
    active={true}
    onClick={[MockFunction]}
    tag="a"
  >
    <div
      className="d-flex p-2 justify-content-center"
    >
      <Icon
        className="mr-1"
        icon="close"
      />
      test
    </div>
  </NavLink>
</NavItem>
`;

exports[`Component: Tab ui without icon: Component: Tab => ui => without icon 1`] = `
<NavItem
  className=""
  tag="li"
>
  <NavLink
    active={true}
    onClick={[MockFunction]}
    tag="a"
  >
    <div
      className="d-flex p-2 justify-content-center"
    >
      test
    </div>
  </NavLink>
</NavItem>
`;
